test/precompilation_test.dart

```diff
@@ -0,0 +1,104 @@
+import 'dart:io';
+
+import 'package:checks/checks.dart';
+import 'package:pub/src/dart.dart';
+import 'package:pub/src/exceptions.dart';
+import 'package:pub/src/log.dart';
+import 'package:test/test.dart';
+
+import 'descriptor.dart';
+
+String outputPath() => '$sandbox/output/snapshot';
+String incrementalDillPath() => '${outputPath()}.incremental';
+
+// A quite big program is needed for the caching to be an actual advantage.
+FileDescriptor foo = file('foo.dart', '''
+foo() {
+  ${List.generate(500000, (index) => 'print("$index");').join('\n')}
+}
+  ''');
+
+FileDescriptor workingMain = file(
+  'main.dart',
+  '''
+import 'foo.dart';
+
+main() async {
+  foo();
+}
+''',
+);
+
+FileDescriptor brokenMain = file(
+  'main.dart',
+  '''
+import 'foo.dart';
+yadda yadda
+main() asyncc {
+  foo();
+}
+''',
+);
+
+Future<Duration> timeCompilation(
+  String executable, {
+  bool fails = false,
+}) async {
+  final s = Stopwatch()..start();
+  verbosity = Verbosity.none;
+  Future<void> compile() async {
+    await precompile(
+      executablePath: executable,
+      name: 'abc',
+      outputPath: outputPath(),
+      packageConfigPath: path('app/.dart_tool/package_config.json'),
+    );
+  }
+
+  if (fails) {
+    await check(compile()).throws<ApplicationException>();
+  } else {
+    await compile();
+  }
+  verbosity = Verbosity.normal;
+  return s.elapsed;
+}
+
+void main() {
+  test('Precompilation is much faster second time and removes old artifacts',
```

Comment:
can we please add a few comments? or at-least a few empty lines :rofl: 

I wouldn't honestly, be okay not having any tests for this. This easily becomes fragile.